src/framework/MOM_diag_buffers.F90

```diff
@@ -0,0 +1,432 @@
+!> Provides buffers that can dynamically grow as needed. These are primarily intended for the
+!! diagnostics which need to store intermediate or partial states of state variables
+module MOM_diag_buffers
+
+use iso_fortran_env, only : stdout => output_unit, stderr => error_unit
+
+! This file is part of MOM6. See LICENSE.md for the license.
+
+implicit none ; private
+
+public :: diag_buffer_unit_tests_2d, diag_buffer_unit_tests_3d
+
+!> The base class for the diagnostic buffers in this module
+type, abstract :: diag_buffer_base ; private
+  integer :: is !< The start slot of the array i-direction
+  integer :: js !< The start slot of the array j-direction
+  integer :: ie !< The end slot of the array i-direction
+  integer :: je !< The end slot of the array j-direction
+  real :: fill_value = 0. !< Set the fill value to use when growing the buffer
```

Comment:
There need to be units ("[arbitrary]") in the comment describing `diag_buffer_base%fill_value`.